clients/pkg/promtail/targets/docker/target.go

```diff
@@ -149,81 +147,98 @@ func (t *Target) processLoop(ctx context.Context) {
 func extractTs(line string) (time.Time, string, error) {
 	pair := strings.SplitN(line, " ", 2)
 	if len(pair) != 2 {
-		return time.Now(), line, fmt.Errorf("Could not find timestamp in '%s'", line)
+		return time.Now(), line, fmt.Errorf("could not find timestamp in '%s'", line)
 	}
 	ts, err := time.Parse("2006-01-02T15:04:05.999999999Z07:00", pair[0])
 	if err != nil {
-		return time.Now(), line, fmt.Errorf("Could not parse timestamp from '%s': %w", pair[0], err)
+		return time.Now(), line, fmt.Errorf("could not parse timestamp from '%s': %w", pair[0], err)
 	}
 	return ts, pair[1], nil
 }
 
-// https://devmarkpro.com/working-big-files-golang
-func readLine(r *bufio.Reader) (string, error) {
-	var (
-		isPrefix = true
-		err      error
-		line, ln []byte
-	)
-
-	for isPrefix && err == nil {
-		line, isPrefix, err = r.ReadLine()
-		ln = append(ln, line...)
-	}
-
-	return string(ln), err
-}
-
-func (t *Target) process(r io.Reader, logStream string) {
+func (t *Target) process(frames chan []byte, logStream string) {
 	defer func() {
 		t.wg.Done()
 	}()
 
-	reader := bufio.NewReader(r)
-	for {
-		line, err := readLine(reader)
+	// softMaxBuffer is to prevent unlimited memory growth.
+	// In principle this introduces the same issue that we are compensating for,
+	// i.e. split messages, however there should be some limit and so this number
+	// is chosen to be "very high" just not "OOM high" (which admittedly is not
+	// a universally defined threshold).
+	const softMaxBuffer = 16 * 1024 * 1024
```

Comment:
if we're using the same 16k as before, how do your changes in this PR get end users a different outcome?

will a log line just be split into multiple log lines instead of being "corrupted" as mentioned in the issue you opened?

Reply:
This is 16 MiB, not 16 kiB :) This message size would likely get rejected at e.g. Loki or other destination anyway. But yes, the proposed behaviour when size goes above 16 MiB is to split the message. I think the only alternative is unbounded buffer growth, which would also seem undesirable, but i can certainly change that if you prefer?

Reply:
Ah yes, you're right :facepalm: 

What if we set this to the default `max_line_size` limit for loki of 256kb? https://grafana.com/docs/loki/latest/configure/#limits_config

We definitely shouldn't have an unbounded buffer, but splitting messages more regularly into pieces that Loki _should_ actually accept would be best.

Reply:
That sounds like a good idea actually. Do you think it should be made a setting then, or should we just pick 256k as the sensible number everyone should use?

I guess this opens up also some kind of hypothetical `max_line_size_exceeded_action` setting which could be e.g. `split`, `truncate`, `discard` (Loki seems to have only a boolean switch for truncate/discard).

Reply:
Ah, I'm not sure how I missed that setting already existing. But I guess that, since this configuration is already there, we should probably follow it instead of hardcoding something? Even though the default seems to actually be unbounded line size...

Reply:
It's now changed to use the configured `max_line_size` although if it is unlimited (set to 0), then we use a max of 256k instead.